Added support for Gitlab

Preeti9764 opened the pull request:
### 📌 Fixes

Fixes #119
---

### 📝 Summary of Changes

- Now the extension also supports Gitlab, and users can generate reports for Gitlab activities just as Github.

![image](https://github.com/user-attachments/assets/978957a8-6c97-4b85-918d-2058a579e12d)




## Summary by Sourcery

Enable the extension to generate Scrum reports from either GitHub or GitLab by adding a platform toggle and integrating GitLab’s API to fetch projects, issues, and merge requests.

New Features:
- Add a platform selector in the popup to switch between GitHub and GitLab
- Persist the selected platform and GitLab username in Chrome storage
- Implement GitLab API integration to fetch user projects, issues, and merge requests

Enhancements:
- Refactor scrumHelper.js to branch data fetching and processing logic by platform
- Introduce ISO date helper functions for query ranges

Preeti9764 commented:
@hpdang @vedansh-5, can you please review it for now? It is currently fetching all the GitLab activities of your public repositories, and I would appreciate your suggestions on this.

vedansh-5 commented:
@Preeti9764 I think it would make more sense to move the organization/group name input box on top with the githubusername and project name.
![image](https://github.com/user-attachments/assets/87b0c54c-7e7e-4a71-91fb-f23610f2f615)
As it would make more sense for the user to set their organization there if someone is not referring to the scrum helpers' readme file.
I think we can give a dropdown list for choosing the client (github/gitlab) instead of the radio, as it would allow us to add more clients without changing the ui in future and would take up less space.

hpdang commented:
@Preeti9764 yes we should have an option to allow users to select their organization, and it can go under Setting. By default if user doesn't select an organization, scrum helper will pull all git activities of user across all organizations? 

Also could you please incorporate the latest changes with the setting button into this PR?

hpdang commented:
> > @Preeti9764 yes we should have an option to allow users to select their organization, and it can go under Setting. By default if user doesn't select an organization, scrum helper will pull all git activities of user across all organizations?
> > Also could you please incorporate the latest changes with the setting button into this PR?
> 
> yes, i will add organisation button in setting which is in are latest ui and set to fetch all activites as default if any organisation is not selected. @hpdang do we need this default for both github as well as gitlab cause for doing it for github i need to make changes in api calling of github .

what is our current logic with github, we don't have to select any org at the moment, it checks FOSSASIA only?

vedansh-5 commented:
> > what is our current logic with github, we don't have to select any org at the moment, it checks FOSSASIA only?
> 
> yes , the default org is fossasia, we can change it to other organisation by making little changes in code which is also added in readme.

@hpdang This would be a nice addition along with org input. We can fetch all the user data within a date range when no org name is set.

Preeti9764 commented:
> > This org name input feature can be tackled in a new PR. @hpdang please share your views on this. thanks
> 
> @vedansh-5 i have made changes for the same and can push it with this PR. I have also seen some changes since last pr#133 is megred the false is again seen in the subject and scrum  subject is not loaded in email and yahoo. can you please check that . Thanks!

vedansh-5 commented:
> > This org name input feature can be tackled in a new PR. @hpdang please share your views on this. thanks
> 
> @vedansh-5 i have made changes for the same and can push it with this PR.

@Preeti9764 I understand you've made code changes to implement the same, which is okay for this PR we wouldn't want to undo your work, ffr-please create another PR if you solve an existing bug or add a new feature, this makes it easier to keep track of the code changes and fixes also its easier to rebase our existing PRs when a new PR is merged, Thanks.

vedansh-5 commented:
> the false is again seen in the subject and scrum subject is not loaded in email and yahoo. can you please check that . Thanks!

This probably will be because of rebasing, I'll look into it. Thanks for pointing out!

hpdang commented:
We should not introduce too many changes in one PR. Let’s break down the changes and address them one by one:
- Issue 1: Enable fetching of all user data within a selected date range when no organization name is set, as we want to support usage beyond FOSSASIA.
- Issue 2: Add functionality to allow users to select an organization in Setting 
- Issue 3: Implement GitLab integration

What do you think? @Preeti9764 @vedansh-5

Preeti9764 commented:
> We should not introduce too many changes in one PR. Let’s break down the changes and address them one by one:
> 
> * Issue 1: Enable fetching of all user data within a selected date range when no organization name is set, as we want to support usage beyond FOSSASIA.
> * Issue 2: Add functionality to allow users to select an organization in Setting
> * Issue 3: Implement GitLab integration
> 
> What do you think? @Preeti9764 @vedansh-5

yes, this could be done . I think this current pr works fine for issue 3. in this all the GitLab activities are fetched for your public repositories.

Preeti9764 commented:
@hpdang Check this pr once i have resolved the confilicts and checked it. It is working fine for me and ready to merge.

vedansh-5 reviewed (approved):
nit: Too much space between the first header and injected content, please fix that.
![image](https://github.com/user-attachments/assets/db9e8c8c-3597-4fe4-b28f-ea5da402e07a)
Please change the github gitlab checkbox to a drop down list. 
Injects content properly in all email clients, few changes and ready to merge.
nit: consider moving all initializations at one place.

Preeti9764 commented:
> nit: Too much space between the first header and injected content, please fix that. 

okay i will solve the gap issue

> Please change the github gitlab checkbox to a drop down list. Injects content properly in all email clients, few changes and ready to merge. 
I dp not think dropdown make sense for two plateforms, we can make that change as we add support of different plateforms.Thanks!

vedansh-5 commented:
![image](https://github.com/user-attachments/assets/8ed94003-bf46-41b7-9f2a-f575df9792d3)

Preeti9764 commented:
@hpdang made the ui changes , please review it and give your feedback whenever you get time. Thanks!

Preeti9764 commented:
![Screenshot 2025-07-10 012934](https://github.com/user-attachments/assets/8e577b97-fd07-47e5-a848-7f8abfa43ea0)
![image](https://github.com/user-attachments/assets/98c0c36b-f09a-4da1-8137-7fa31b3cba95)

Preeti9764 commented:
> @hpdang Made the above changes , for the fresh  load unpacked the plateform is selected to github and will be saved as the user switches it, Is that okay ? Thanks! ![image](https://private-user-images.githubusercontent.com/193853272/464671477-3cc802bd-fe94-468e-bb60-cdf13ef53d0a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTIxNDUwMzAsIm5iZiI6MTc1MjE0NDczMCwicGF0aCI6Ii8xOTM4NTMyNzIvNDY0NjcxNDc3LTNjYzgwMmJkLWZlOTQtNDY4ZS1iYjYwLWNkZjEzZWY1M2QwYS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNzEwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDcxMFQxMDUyMTBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kZmNkMTFjOTE5YmU1N2Q4MDlkMjYwZmRmMDQ3MjMyNWVlMGQ5MzQzNDgwZTc2NjVmYWIzN2Y3ZTM4NDk4YTk2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.eXcAhJVglD9-Md08N6BoGgmGz5v5Z1XP1JoK-D8ZHtI)

Preeti9764 commented:
@hpdang  I have encountered a bug and i am working on that , please suggest your view for token , should i also add the gitlab token functionality just as github tokens? or should i first add the basic gitlab support cause replicate it as github , it has many things like org name are not there in gitlab but groups are there .Please guide regarding this. Thanks!

hpdang commented:
@Preeti9764 Let's try to get the basic functionality work first in this PR. Let's do the token and other setting items in the next PR.

Preeti9764 commented:
> @Preeti9764 Let's try to get the basic functionality work first in this PR and do the token and other setting items in the next PR.

alright!

Preeti9764 commented:
@hpdang @vedansh-5 made all the changes, working fine from my side , please review .thanks!

hpdang commented:
@Preeti9764 When GitLab is selected, the Generate Report screen displays the correct content, but the Compose window still shows content from GitHub

Preeti9764 commented:
> @Preeti9764 When GitLab is selected, the Generate Report screen displays the correct content, but the Compose window still shows content from GitHub

@hpdang for me it works fine now .
<img width="683" height="647" alt="image" src="https://github.com/user-attachments/assets/02b2d9ad-3dda-43c4-8ad3-46a36b157f21" />

hpdang commented:
@vedansh-5 please test the latest changes

vedansh-5 commented:
@Preeti9764 I can not enable the extension. Also the enable disable toggles' state is not being saved. 
I can not click on any of the fields.
The settings button does not work. 
In dark mode the drop down list button is still grey.
The drop down list is see through because of which the help text of project name is visible completely.
Thankyou

https://github.com/user-attachments/assets/63c9fbcb-ce5e-4345-97db-313f75afa37f

vedansh-5 commented:
> @vedansh-5 I cannot reproduce this , can you please check with the latest commits .Thanks!

<img width="1919" height="1079" alt="image" src="https://github.com/user-attachments/assets/1e20d23b-8957-41b8-b427-539cfe7f5848" />
I had the latest changes. I pulled again and reloaded the extension. Same thing even now.

vedansh-5 commented:
`tailwindcss:1  Failed to load resource: net::ERR_FILE_NOT_FOUND
emailClientAdapter.js:216 Email client adapter initialized
main.js:259 Uncaught TypeError: Cannot read properties of null (reading 'addEventListener')
    at main.js:259:23
scrumHelper.js:13 Script loaded, adapter exists: true
popup.js:442 Switched to report view
popup.js:442 Switched to report view
popup.js:442 Switched to report view
popup.js:104 Uncaught ReferenceError: tokenPreview is not defined
    at renderTokenPreview (popup.js:104:9)
    at HTMLImageElement.<anonymous> (popup.js:100:9)
renderTokenPreview @ popup.js:104
(anonymous) @ popup.js:100
popup.js:104 Uncaught ReferenceError: tokenPreview is not defined
    at renderTokenPreview (popup.js:104:9)
    at HTMLImageElement.<anonymous> (popup.js:100:9)
renderTokenPreview @ popup.js:104
(anonymous) @ popup.js:100
`
These are my console logs

Preeti9764 commented:
@vedansh-5 Fixed the disable bug , Please review Thanks!

Preeti9764 commented:
@hpdang Made the required changes, ready from my end . Please review.Thanks

vedansh-5 commented:
<img width="468" height="750" alt="image" src="https://github.com/user-attachments/assets/30a701db-e48a-43ff-b7b4-2c968ff4e742" />
Please fix the inconsistent color in dark mode.

<img width="471" height="753" alt="image" src="https://github.com/user-attachments/assets/394f3247-e298-40e8-a342-868f55952ccb" />
Its not fetching the updates for my gitlab,
<img width="1360" height="392" alt="image" src="https://github.com/user-attachments/assets/a1cb238f-bc79-4894-a06e-bbfefd73bfbd" />
Also the state of repo filter is not being retained, it is checked everytime I reopen the extension window.
@Preeti9764 Please take a look at these, thanks!

vedansh-5 commented:
Also I'm not sure if the show commits feature work in gitlabs, so change the helper text for gitlab context stating the same.
<img width="423" height="236" alt="image" src="https://github.com/user-attachments/assets/c79b2b1e-4641-4ec3-9f24-5b3fb9543d43" />

The help bubble is visible from under the list, that shouldn't be the case, for this make the dropdown list a little wider and set its z index to more than that of help text. See repo dropdown list, essentially set the z-index same to it.

Preeti9764 commented:
@hpdang done with changes ,please check and give your views .Thanks!

Preeti9764 commented:
Your vedansh-project is private project , its activites cannot be fetched through public api request.Thanks!

Preeti9764 commented:
@hpdang @vedansh-5 ready for review.Thanks!

Preeti9764 commented:
@hpdang @vedansh-5 ready for review.Thanks!

vedansh-5 commented:
> Your vedansh-project is private project , its activites cannot be fetched through public api request.Thanks!

@Preeti9764 I do not intend to fetch the activities of `vedansh-project`. I want to fetch the activities for `scrumHelperTest` which is a public project inside a public group.
https://gitlab.com/vedansh-group/scrumhelpertest.
I still am unable to fetch for it. 
<img width="469" height="751" alt="image" src="https://github.com/user-attachments/assets/82b9841f-78d0-40b3-b0e3-5ae178206252" />
Please go through the console logs, ping me if you need something else
<img width="806" height="718" alt="image" src="https://github.com/user-attachments/assets/ca0f34ec-96bb-49cb-81f5-518a432e5ace" />
Thanks!

Preeti9764 commented:
@vedansh-5 correctly fetching your gitlab actvities.Thanks!
<img width="427" height="722" alt="image" src="https://github.com/user-attachments/assets/9ee8651b-6106-4f6d-af51-187321847bcb" />

vedansh-5 commented:
@Preeti9764 I observed a bug whenever I pressed refresh data button, `[object Object]` appeared at the end of the extension. To confirm whether it was branch specific or is the same issue in the master branch too, I removed the extension and loaded the master branch to check. I did not see the same issue there, then loaded your feature branch again and the issue didn't come up this time. Fixing of bugs/issues on their own is a strong indicator towards race condition. Please check the codebase for that. Ott this PR lgtm! Thankyou.

vedansh-5 commented:
> 
> > Despite choosing an org all the data is being fetched.
> 
> I think you mean repo not a org cause org is empty and it will fetch all activities?

Yes my bad, I mean repo.

Preeti9764 commented:
@vedansh-5 @hpdang Please review.Thanks!

Preeti9764 commented:
> it works for me. @Preeti9764 could you please add a space between Platform and the grey box, make it similar to Project Name box
@hpdang I have added the space , it is ready to merge .Thanks!
